### PR TITLE
[SPARK-42406] Terminate Protobuf recursive fields by dropping the field

### PR DESCRIPTION
**Pull request by rangadi**
### What changes were proposed in this pull request?

Protobuf deserializer (`from_protobuf()` function()) optionally supports recursive fields up to certain depth. Currently it uses `NullType` to terminate the recursion. But an `ArrayType` containing `NullType` is not really useful and  it does not work delta.

This PR fixes this by removing the field to terminate recursion rather than using `NullType`. 
The following example illustrates the difference. 

E.g. Consider a recursive Protobuf like this:
```
message Node {
    int value = 1;
    repeated Node children = 2  // recursive array
}
message Tree {
    Node root = 1
}
```
Catalyst schama with `from_protobuf()` of `Tree` with max recursive depth set to 2, would be:
 
   - **Before**:  _STRUCT<root: STRUCT<value: int, children: array<STRUCT<value: int, **children: array< void >**>>>>_
   - **After**: _STRUCT<root: STRUCT<value: int, children: array<STRUCT<value: int>>>>_

Notice that at second level, the `children` array is dropped, rather than being defined as `array<void>`. 

### Why are the changes needed?
 - This improves how Protobuf connector handles recursive fields. It avoids using `void` fields which are problematic in many scenarios and do not add any information.

### Does this PR introduce _any_ user-facing change?
 - This changes the schema in a subtle manner while using with recursive support enabled. Since this only removes an optional field, it is backward compatible. 

### How was this patch tested?
 - Added multiple unit tests and updated existing one. Most of the changes for this PR are in the tests.

### CODE REVIEW

File: connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/ProtobufDeserializer.scala [hunk 1]
```diff
@@ -157,9 +157,6 @@ private[sql] class ProtobufDeserializer(
     (protoType.getJavaType, catalystType) match {
 
       case (null, NullType) => (updater, ordinal, _) => updater.setNullAt(ordinal)
-      // It is possible that this will result in data being dropped, This is intentional,
-      // to catch recursive fields and drop them as necessary.
-      case (MESSAGE, NullType) => (updater, ordinal, _) => updater.setNullAt(ordinal)
```
Comment (rangadi) [inline connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/ProtobufDeserializer.scala:162]: `NullType` is no longer expected.

File: connector/protobuf/src/test/scala/org/apache/spark/sql/protobuf/ProtobufFunctionsSuite.scala [hunk 2]
```diff
@@ -1101,166 +939,61 @@ class ProtobufFunctionsSuite extends QueryTest with SharedSparkSession with Prot
...
-         |            "nullable" : true
-         |          } ]
-         |        },
-         |        "nullable" : true
-         |      } ]
-         |    },
-         |    "nullable" : true
-         |  } ]
-         |}""".stripMargin).asInstanceOf[StructType]
+    val schemaOne = structFromDDL(
```
Comment (rangadi) [inline connector/protobuf/src/test/scala/org/apache/spark/sql/protobuf/ProtobufFunctionsSuite.scala:942]: As part of this PR, replaced many long json schemas with DDL which is lot more concise.

File: connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/utils/SchemaConverters.scala [hunk 3]
```diff
@@ -59,6 +58,8 @@ object SchemaConverters {
   // existingRecordNames: Map[String, Int] used to track the depth of recursive fields and to
   // ensure that the conversion of the protobuf message to a Spark SQL StructType object does not
   // exceed the maximum recursive depth specified by the recursiveFieldMaxDepth option.
+  // A return of None implies the field has reached the maximum allowed recursive depth and
```
Comment (rangadi) [inline connector/protobuf/src/main/scala/org/apache/spark/sql/protobuf/utils/SchemaConverters.scala:63]: This file has the main fixes. A 'None' is returned when recursion depth is reached. And the field is dropped by the caller (i.e. outer struct).

**Review by gengliangwang — APPROVED** [review]

**Comment by gengliangwang** [comment]
+1, the new schema is more reasonable.

**Comment by gengliangwang** [comment]
Thanks, merging to master/3.4

**Pull request closed without merging**